backend/remote-state/swift/backend_state.go

```diff
@@ -1,22 +1,91 @@
 package swift
 
 import (
+	"fmt"
+	"strings"
+
 	"github.com/hashicorp/terraform/backend"
 	"github.com/hashicorp/terraform/state"
 	"github.com/hashicorp/terraform/state/remote"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+const (
+	objectEnvPrefix = "env-"
+	delimiter       = "/"
 )
 
 func (b *Backend) States() ([]string, error) {
-	return nil, backend.ErrNamedStatesNotSupported
+	client := &RemoteClient{
+		client:           b.client,
+		container:        b.container,
+		archive:          b.archive,
+		archiveContainer: b.archiveContainer,
+		expireSecs:       b.expireSecs,
+	}
+
+	// List our container objects
+	objectNames, err := client.ListObjectsNames(objectEnvPrefix, delimiter)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the envs, we use a map since we can get duplicates with
+	// path suffixes.
+	envs := map[string]struct{}{}
+
+	for _, object := range objectNames {
+		object = strings.TrimPrefix(object, objectEnvPrefix)
+		object = strings.TrimSuffix(object, delimiter)
+
+		// Ignore objects that still contain a "/"
+		// as we dont store states in subdirectories
+		if idx := strings.Index(object, delimiter); idx >= 0 {
+			continue
+		}
+
+		envs[object] = struct{}{}
+	}
+
+	if err != nil {
+		return nil, err
+	}
```

Comment:
From what I can tell, there's no new error being generated between here and line 30, so can this be removed?